plugins/ltac2/tac2ffi.ml

```diff
@@ -476,3 +476,8 @@ let app_fun1 cls r0 r1 x =
 let annotate_closure fr (MLTactic (arity, fr0, f)) =
   assert (Option.is_empty fr0);
   MLTactic (arity, Some fr, f)
+
+let rec purify_closure : type a. a arity -> a -> a = fun arity f ->
```

Comment:
This does not fully prevent side effects from leaking from the closure. If the library author is really careless, he could write something like `fun x -> let () = nasty_effect () in fun y -> body` and AFAICT this code is not going to properly wrap it.

Reply:
Do we care? It's not easy to deal with the gadt stuff to wrap every sub application.

Reply:
Then I'd rather fix the functions performing side-effects in Tac2core with an explicit caveat in the doc, rather than adding an imperfect wrapper that only adds the illusion of safety.